# biomed-multi-omic

Biomedical foundational models for omics data. This package supports the development of foundation models for scRNA or for DNA data.


- ðŸ§¬ A single package for DNA and RNA Foundation models
- ðŸš€ Built around anndata, HuggingFace transformers and PyTorchLighting to integrate into existing software stacks
- ðŸ“ˆ Increased performance on AvgBio scores using Whole Cell Expression Decoder (WCED)
- ðŸ”¬ Most expansive package for Transcriptomic Foundation Models (TPM) and Genomics Foundation Models

## Contents

1. [Introduction](#introduction)
2. [Installation](#installation)
3. [Quick start](#quick-start)
4. [Citation](#citation)

## Introduction

`biomed-multi-omic` is a comprehensive software package that enables  exploration of diverse multiomic foundation model methods, including incorporating novel methods as the field continues to advance. It also facilitates downstream tasks such as cell type annotation, perturbation prediction, and batch effect
correction on benchmark datasets. `biomed-multi-omic` is built upon anndata, HuggingFace Transformers and PyTorchLighting to ensure integration into existing software stacks.

## Package Architecture

### RNA Modules

`bmfm-rna` framework diagram schematic shows the modules available for building  Transcriptomics Foundation Model (TPM). A novel contribution of our work is the Whole Cell Expression Decoder (WCED), an innovative pretraining method aimed at improving transcriptomic foundation models. In WCED, the modelâ€™s objective is to reconstruct a full cell expression profile from a partial input's `[CLS]` token representation generated by the transformer encoder. By training models to autocomplete the expression profiles, WCED improves the modelâ€™s understanding of underlying biological processes, resulting in better generalization and more accurate predictions for downstream tasks.

![bmfm_omics_workflow](docs/images/package_diagram.png)

## DNA Modules

The `bmfm-dna` framework addresses key limitations of existing DNA language models by incorporating natural genomic variations into the pre-training process, rather than relying solely on the reference genome. This allows the model to better capture critical biological properties, especially in regulatory regions where many disease-associated variants reside.

![bmfm_dna](docs/images/dna_fig1.png)

`bmfm-dna` framework diagram schematic shows the modules available for multiple strategies to encode natural genomic variations; fine-tuning and benchmarking of the pre-trained models on well-established biologically meaningful tasks, such as Evaluation (GUE) package released in DNABERT-2, promoter activity prediction on Massive Parallel Reporting Assays (MPRA) dataset, and SNP-disease association prediction.

For more details, check out the [the BMFM-DNA preprint](https://www.arxiv.org/abs/2507.05265)..

## Installation

We recommend using [uv](https://github.com/astral-sh/uv) to create your environment due to it's 10-100x speed up over pip.


Install using cloned repo:

```sh
git clone git@github.com:BiomedSciAI/biomed-multi-omic.git
cd biomed-multi-omics
uv venv .venv -p3.12
source ./.venv/bin/activate
uv pip install -e .
```

NB - `biomed-multi-omic` depends on `hic-straw` which requires `curl`. You may need to install `curl` or `libcurl` , for more information please refer to `curl`'s install instructions for your OS.

### Optional dependencies

In addition to the base package there are additional optional dependencies which extends `bmfm-mulit-omics` capabilites further. These include:

- `bulk_rna`: Extends modules for extracting and preprocessing bulk RNA-seq data
- `benchmarking`: Installs additional models used benchmark `bmfm-mulit-omics` against. These include scib, scib-metrics, pyliger, scanorama and harmony-pytorch.
- `test`: Unittest suite which is recommended for development use

To install optional dependencies from this GitHub repository you can run from package root:

```sh
uv pip install ".[bulk_rna,benchmarking,test,notebook]"
```

## Quick start

### Downloading RNA checkpoint

The model's weights can be aquired from [IBM's HuggingFace collection](https://huggingface.co/ibm-research). The following scRNA models are avaliable:

- MLM+RDA: [ibm-research/biomed.rna.bert.110m.mlm.rda.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.mlm.rda.v1)
- MLM+Multitask: [ibm-research/biomed.rna.bert.110m.mlm.multitask.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.mlm.multitask.v1)
- WCED+Multitask: [ibm-research/biomed.rna.bert.110m.wced.multitask.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.wced.multitask.v1)
- WCED 10 pct: [ibm-research/biomed.rna.bert.110m.wced.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.wced.v1)

For details on how the models were trained, please refer to [the BMFM-RNA preprint](https://arxiv.org/abs/2506.14861).

### RNA Zero-shot inference (CLI)

`biomed-multi-omic` allows for multiple input data types, but the core data object at the heart of the tool is based around the adata object. Once your dataset is created.

To get scRNA embeddings and zero shot cell-type predictions:

```bash
export MY_DATA_FILE=... # path to h5ad file with raw counts and gene symbols
bmfm-targets-run -cn predict input_file=$MY_DATA_FILE working_dir=/tmp checkpoint=ibm-research/biomed.rna.bert.110m.wced.multitask.v1
```

For more details tutorial see RNA tutoirals. Note to use the notebook you will need to install the `notebook` optional dependencies (see [Installation](#2-installation)):

- To run inference programmatically, you can see a zero-shot example in this [scRNA zero-shot notebook](tutorials/RNA/1_zero_shot_using_yaml.ipynb).
- To inspect the resulting embeddings and cell-type predictions use this [scRNA inspect embeddings notebook](tutorials/RNA/2_inference_inspection.ipynb).

### Downloading DNA checkpoint

The model's weights can be aquired from IBM's HuggingFace collection. The following DNA models are avaliable:

- MLM+REF_GENOME: [ibm-research/biomed.dna.ref.modernbert.113m](https://huggingface.co/ibm-research/biomed.dna.ref.modernbert.113m.v1)
- MLM+REFSNP_GENOME: [ibm-research/biomed.dna.snp.modernbert.113m](https://huggingface.co/ibm-research/biomed.dna.snp.modernbert.113m.v1)

### DNA Inference

For details on how the models were trained, please refer to the BMFM-DNA preprint.

To perform DNA zero-shot:

```bash
export MY_DATA_FILE=... # path to your DNA sequences files
bmfm-targets-run -cn dna_predict input_file=$MY_DATA_FILE working_dir=/tmp checkpoint=ibm-research/biomed.dna.snp.modernbert.113m.v1
```

For more details tutorial see DNA tutoirals.

## Citation

To cite the tool for both RNA and DNA, please cite both the following articles:

```bibtex
@misc{dandala2025bmfmrnaopenframeworkbuilding,
      title={BMFM-RNA: An Open Framework for Building and Evaluating Transcriptomic Foundation Models},
      author={Bharath Dandala and Michael M. Danziger and Ella Barkan and Tanwi Biswas and Viatcheslav Gurev and Jianying Hu and Matthew Madgwick and Akira Koseki and Tal Kozlovski and Michal Rosen-Zvi and Yishai Shimoni and Ching-Huei Tsou},
      year={2025},
      eprint={2506.14861},
      archivePrefix={arXiv},
      primaryClass={q-bio.GN},
      url={https://arxiv.org/abs/2506.14861},
}

@misc{li2025bmfmdnasnpawarednafoundation,
      title={BMFM-DNA: A SNP-aware DNA foundation model to capture variant effects},
      author={Hongyang Li and Sanjoy Dey and Bum Chul Kwon and Michael Danziger and Michal Rosen-Tzvi and Jianying Hu and James Kozloski and Ching-Huei Tsou and Bharath Dandala and Pablo Meyer},
      year={2025},
      eprint={2507.05265},
      archivePrefix={arXiv},
      primaryClass={q-bio.GN},
      url={https://arxiv.org/abs/2507.05265},
}
```
